backend/zkppSalt.js

```diff
@@ -0,0 +1,229 @@
+  }
+
+  const encryptedSecretKeyBuf = Buffer.from(encryptedSecretKey.replace(/_/g, '/').replace(/-/g, '+'), 'base64')
+  const encryptionKey = hashStringArray('REG', contract, registrationSecret).slice(0, nacl.secretbox.keyLength)
+  const secretKeyBuf = nacl.secretbox.open(encryptedSecretKeyBuf.slice(nacl.secretbox.nonceLength), encryptedSecretKeyBuf.slice(0, nacl.secretbox.nonceLength), encryptionKey)
+
+  const parseRegisterSaltRes = parseRegisterSalt(clientPublicKey, secretKeyBuf, encryptedHashedPassword)
+
+  if (!parseRegisterSaltRes) {
+    return false
+  }
+
+  const [authSalt, contractSalt, hashedPasswordBuf] = parseRegisterSaltRes
+
+  await setZkppSaltRecord(contract, Buffer.from(hashedPasswordBuf).toString(), authSalt, contractSalt)
+
+  return true
+}
+
+const contractSaltVerifyC = (h: string, r: string, s: string, userHc: string) => {
+  const [c, hc] = computeCAndHc(r, s, h)
+  const userHcBuf = Buffer.from(userHc.replace(/_/g, '/').replace(/-/g, '+'), 'base64')
+
+  if (hc.byteLength === userHcBuf.byteLength && timingSafeEqual(hc, userHcBuf)) {
+    return c
+  }
+
+  return false
+}
+
+export const getContractSalt = async (contract: string, r: string, s: string, sig: string, hc: string): Promise<false | string> => {
+  if (!verifyChallenge(contract, r, s, sig)) {
+    return false
+  }
+
+  const record = await getZkppSaltRecord(contract)
+  if (!record) {
+    return false
+  }
+
+  const { hashedPassword, contractSalt } = record
+
+  const c = contractSaltVerifyC(hashedPassword, r, s, hc)
+
+  if (!c) {
+    return false
+  }
+
+  return encryptContractSalt(c, contractSalt)
+}
+
+export const update = async (contract: string, r: string, s: string, sig: string, hc: string, encryptedArgs: string): Promise<boolean> => {
+  if (!verifyChallenge(contract, r, s, sig)) {
+    return false
+  }
+
+  const record = await getZkppSaltRecord(contract)
+  if (!record) {
+    return false
+  }
+  const { hashedPassword } = record
+
+  const c = contractSaltVerifyC(hashedPassword, r, s, hc)
+
+  if (!c) {
+    return false
+  }
+
+  const encryptionKey = hashRawStringArray('SU', c).slice(0, nacl.secretbox.keyLength)
+  const encryptedArgsBuf = Buffer.from(encryptedArgs.replace(/_/g, '/').replace(/-/g, '+'), 'base64')
+  const nonce = encryptedArgsBuf.slice(0, nacl.secretbox.nonceLength)
+  const encrytedArgsCiphertext = encryptedArgsBuf.slice(nacl.secretbox.nonceLength)
+
+  const args = nacl.secretbox.open(encrytedArgsCiphertext, nonce, encryptionKey)
+
+  if (!args) {
+    return false
+  }
+
+  try {
+    const argsObj = JSON.parse(Buffer.from(args).toString())
+
+    if (!Array.isArray(argsObj) || argsObj.length !== 3 || !argsObj.reduce((acc, cv) => acc && typeof cv === 'string', true)) {
+      return false
+    }
+
+    const [hashedPassword, authSalt, contractSalt] = argsObj
+
+    await setZkppSaltRecord(contract, hashedPassword, authSalt, contractSalt)
+
+    return true
+  } catch {
+    // empty
+  }
+
+  return false
```

Comment:
This file overall could use a lot of commentary. Can you please update it so that:

- Each function has a comment above it explaining its purpose
- Anything that can be interpreted as odd, unusual, non-idiomatic, or even simply unclear, should have commentary added explaining the purpose and reasoning behind it
- Various places that have result in errors need corresponding `console.warn` or `console.error` statements. I commented on some but not all, please give the file a look over and add appropriate logging so that if something goes wrong we can track down what exactly it was